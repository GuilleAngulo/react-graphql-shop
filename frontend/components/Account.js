import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';

import Form from './styles/Form';
import Error from './ErrorMessage';


const UPDATE_USER_MUTATION = gql`
    mutation UPDATE_USER_MUTATION(
        $id: ID!
        $name: String
        $email: String
        $password: String
    ) {
    updateUser(id: $id, name: $name, email: $email, password: $password) {
            id
            name
            email
        }
    }
`;


class Account extends Component {
    
    static propTypes = {
        user: PropTypes.shape({
            id: PropTypes.string.isRequired,
            name: PropTypes.string.isRequired,
            email: PropTypes.string.isRequired,
        }).isRequired,
    };

    state = { 
        id: this.props.user.id,
        name: this.props.user.name,
        email: this.props.user.email,
    };

    handleChange = e => {
        const { name, value } = e.target;
        this.setState({ [name]: value });
    }

    updateUser = async (e, updateUserMutation) => {
        e.preventDefault();
        const res = await updateUserMutation({
            variables: {
                ...this.state,
            },
        });
    }

    render() {
        const { user } = this.props;
        return (
                <Mutation 
                    mutation={ UPDATE_USER_MUTATION }
                    variables={this.state}>
                        {(updateUser, { loading, error }) => user && (
                            <Form 
                            data-test="form"
                            onSubmit={e => this.updateUser(e, updateUser)}>
                                <Error error={error} />
                                <fieldset disabled={loading} aria-busy={loading}>
                                    
                                    <label htmlFor="name">
                                        Name
                                        <input 
                                            type="text" 
                                            id="name" 
                                            name="name" 
                                            placeholder="Name" 
                                            required 
                                            defaultValue={user.name}
                                            onChange={this.handleChange}
                                            />
                                    </label>

                                    <label htmlFor="email">
                                        Email
                                        <input 
                                            type="text" 
                                            id="email" 
                                            name="email" 
                                            placeholder="Email" 
                                            required 
                                            defaultValue={user.email}
                                            onChange={this.handleChange}
                                            />
                                    </label>

                                    <label htmlFor="password">
                                        Your Actual Password
                                        <input 
                                            type="password" 
                                            id="password" 
                                            name="password" 
                                            placeholder="Type your current password" 
                                            required
                                            onChange={this.handleChange}
                                            />
                                    </label>
                                    <button type="submit">Sav{loading ? 'ing' : 'e'} Changes</button>
                                </fieldset>
                            </Form>
                        )}    
                    </Mutation>
                
        );
    }
}

export default Account;
export { UPDATE_USER_MUTATION };