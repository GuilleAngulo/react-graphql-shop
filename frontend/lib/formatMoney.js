import { country } from '../config';

export default function(amount) {

  const options = [
    {
      locales: 'en-US',
      option: {
        style: 'currency',
        currency: 'USD',
        minimumFractionDigits: 2,
      },
    },
    {
      locales: 'pt-BR',
      option: {
        style: 'currency',
        currency: 'BRL',
        minimumFractionDigits: 2,
      },
    },
    {
      locales:  'es-ES',
      option: {
        style: 'currency',
        currency: 'EUR',
        minimumFractionDigits: 2,
      },
    },
  ];

  const { locales, option } = options.find(option => option.locales === country);

  // if its a whole, dollar amount, leave off the .00
  if (amount % 100 === 0) option.minimumFractionDigits = 0;

  const formatter = new Intl.NumberFormat(locales, option);
  return formatter.format(amount / 100);
};