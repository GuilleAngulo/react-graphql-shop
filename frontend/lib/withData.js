import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { LOCAL_STATE_QUERY } from '../components/Cart';

import { endpoint } from '../config';

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    //LOCAL DATA
    clientState: {
      resolvers: {
        Mutation: {
          toogleCart(_, variables, { cache }) {
            //Read the cartOpen value from the cache
            const { cartOpen } = cache.readQuery({
              query: LOCAL_STATE_QUERY,
            });
            //Write the cart State to the opposite
            const data = {
              data: { cartOpen: !cartOpen }
            };
            cache.writeData(data);
            return data;
          },
        },
      },
      defaults: {
        cartOpen: false,
      },
    },
  });
}

export default withApollo(createClient);
