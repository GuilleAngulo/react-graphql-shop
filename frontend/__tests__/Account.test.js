import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import { ApolloConsumer} from 'react-apollo';
import Account, { UPDATE_USER_MUTATION } from '../components/Account';
import { fakeUser } from '../lib/testUtils';

const user = fakeUser();
const mocks = [
    {
        request: { query: UPDATE_USER_MUTATION },
        result: { 
            data: {
                me: {
                ...fakeUser(),
                },
            },
        },
    },
];

describe('<Account/>', () => {
    it('renders and matches the snapshot', async () => {
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <Account user={ {...user }}/>
            </MockedProvider>
        );
        await wait(),
        wrapper.update();
        const form = wrapper.find('form[data-test="form"]');
        expect(toJSON(form)).toMatchSnapshot();
    });
});