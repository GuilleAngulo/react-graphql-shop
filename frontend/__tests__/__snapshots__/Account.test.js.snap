// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Account/> renders and matches the snapshot 1`] = `
<form
  className="Form-sc-4ldhx0-0 cTjOt"
  data-test="form"
  onSubmit={[Function]}
>
  <DisplayError
    error={Object {}}
  />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="name"
    >
      Name
      <input
        defaultValue="Miss Coleman Berge"
        id="name"
        name="name"
        onChange={[Function]}
        placeholder="Name"
        required={true}
        type="text"
      />
    </label>
    <label
      htmlFor="email"
    >
      Email
      <input
        defaultValue="Delmer.Smith@yahoo.com"
        id="email"
        name="email"
        onChange={[Function]}
        placeholder="Email"
        required={true}
        type="text"
      />
    </label>
    <label
      htmlFor="password"
    >
      Your Actual Password
      <input
        id="password"
        name="password"
        onChange={[Function]}
        placeholder="Type your current password"
        required={true}
        type="password"
      />
    </label>
    <button
      type="submit"
    >
      Sav
      e
       Changes
    </button>
  </fieldset>
</form>
`;
